pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -18,32 +18,126 @@
+
+    //this is for replay
+    private final PositionImpl lastConfirmedEntry;
+    //this if for replay
+    private PositionImpl currentLoadPosition;
+
+    private final AtomicInteger countToSyncPosition = new AtomicInteger(0);
+
+    private final static String TXN_ON_GOING_POSITION_SUFFIX = "-txnOnGoingPosition";
+
+    private final String txnOnGoingPositionName;
+
+    //TODO this can config
+    private int defaultCountToSyncPosition = 10000;
+
+    public TopicTransactionBuffer(PersistentTopic topic) throws ManagedLedgerException {
+        super(State.None);
+        this.entryQueue = new SpscArrayQueue<>(2000);
         this.topic = topic;
+        ManagedLedger managedLedger = topic.getManagedLedger();
+        this.lastConfirmedEntry = (PositionImpl) managedLedger.getLastConfirmedEntry();
+        this.txnOnGoingPositionName = topic.getName() + TXN_ON_GOING_POSITION_SUFFIX;
+        String positionString = managedLedger.getProperties().get(txnOnGoingPositionName);
+        if (positionString == null) {
+            this.currentLoadPosition = PositionImpl.earliest;
+        } else {
+            PositionImpl position = PositionImpl.earliest;
+            try {
+                position = PositionImpl.convertStringToPosition(positionString);
+            } catch (Exception e) {
+                log.error("Topic : [{}] transaction buffer get replay start position error!", topic.getName());
+            }
+            this.currentLoadPosition = position;
+        }
+        this.cursor = managedLedger.newNonDurableCursor(currentLoadPosition);
+
+        new Thread(() -> new TopicTransactionBufferReplayer(new TransactionBufferReplayCallback() {
+
+            @Override
+            public void replayComplete() {
+                if (!changeToReadyState()) {
+                    log.error("Managed ledger transaction metadata store change state error when replay complete");
+                }
+            }
+
+            @Override
+            public void handleMetadataEntry(Position position, MessageMetadata messageMetadata) {
+                if (!messageMetadata.hasTxnidMostBits() || !messageMetadata.hasTxnidLeastBits()) {
+                    return;
+                }
+                TxnID txnID = new TxnID(messageMetadata.getTxnidMostBits(),
+                        messageMetadata.getTxnidLeastBits());
+                if (isTxnMarker(messageMetadata)) {
+                    ConcurrentOpenHashSet<PositionImpl> positions = txnBufferCache.remove(txnID);
+                    positionsSort.removeAll(positions.values());
+                } else {
+                    ConcurrentOpenHashSet<PositionImpl> positions =
+                            txnBufferCache.computeIfAbsent(txnID, (v) -> new ConcurrentOpenHashSet<>());
+                    positions.add((PositionImpl) position);
+                    positionsSort.add((PositionImpl) position);
+                }
+            }
+        }).start()).start();
```

Comment:
we should not start a Thread in a constructor, especially when the Thread refers to the object that is beeing created, we could have bad memory leaks in case we are not properly releasing all of the resources

Reply:
it is good idea, i will change it. :)